grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+            return Ok(StageResult::new(
+                StageStatus::UnrecoverableFailure,
+                Some(Output::new().section(sect)),
+            ));
+        }
+
+        match res.status {
+            ExitStatus::Ok => {
+                run_updates.add_update(Update::new_pass("running valgrind"));
+                results_sect.add_content(log);
+                sect.add_content(run_updates);
+                sect.add_content(Content::SubSection(results_sect));
+
+                Ok(StageResult::new(
+                    StageStatus::Continue {
+                        points_lost: PointQuantity::zero(),
+                    },
+                    Some(Output::new().section(sect)),
+                ))
+            }
+            ExitStatus::Failure(rc) => {
+                if rc < ERROR_EXITCODE {
+                    return Err(anyhow!(
+                        "Error: expected exit code to be at least {}, was {}",
+                        ERROR_EXITCODE,
+                        rc
+                    ));
+                }
+
+                run_updates.add_update(
+                    Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                        .notes("Valgrind Errors Detected"),
+                );
+                results_sect.add_content(log);
+                sect.add_content(run_updates);
+                sect.add_content(Content::SubSection(results_sect));
+
+                return Ok(StageResult::new(
+                    StageStatus::UnrecoverableFailure,
+                    Some(Output::new().section(sect)),
+                ));
+            }
+            _ => panic!("Expected either Ok or Failure ExitStatus if the command completed"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        path::PathBuf,
+        sync::{Arc, Mutex},
+    };
+
+    use genos::{
+        formatter::MarkdownFormatter,
+        output::Contains,
+        process,
+        test_util::{MockDir, MockExecutorInner, MockProcessExecutor},
+        writer::Transform,
+    };
+
+    use super::*;
+
+    use tokio::sync::OnceCell;
+
+    #[derive(Deserialize)]
+    struct ExamplesPrePost {
+        pre: String,
+        post: String,
+    }
+
+    #[derive(Deserialize)]
+    struct ExamplesDebugRelease<T> {
+        debug: T,
+        release: T,
+    }
+
+    #[derive(Deserialize)]
+    struct Examples {
+        noop: String,
+        infinite: ExamplesPrePost,
+        segfault: ExamplesPrePost,
+        agony: ExamplesDebugRelease<ExamplesPrePost>,
+    }
+
+    impl ValgrindConfig {
+        fn new(log_file: &str) -> Self {
+            Self {
+                log_file: log_file.to_string(),
+                ..Self::default()
+            }
+        }
+    }
+
+    // reads various valgrind output examples from YAML
+    // use get_examples to access
+    const EXAMPLES_PATH: &'static str = "resources/valgrind/examples.yml";
+    static EXAMPLES: OnceCell<Examples> = OnceCell::const_new();
```

Comment:
Just so you know, There's a crate called lazy_static, which is doing the same thing you're doing here which may have been useful. It's a pretty commonly used crate. Feel free to use or keep this.

Reply:
Thanks! I actually found info about that first, but didn't want to add a dependency if possible. From what I remember the syntax is not much more complicated in this case, so I'll just leave it

Reply:
sounds good!